<b>Import gurobipy module to perform optimization for linear programming problems</b>

In [1]:
from gurobipy import *

<b>Part 1: Value Com Electronics Production Maximization Problem</b>

<p>Problem description: Value Com manufactures five different models of NIC for<br>
Pcs and Desktops. Each NIC models requires specific amount of PCBs, resistors,<br>
Memory chips and labor hours. We have to achieve desired production of minimum<br>
500 units of each model while maximizing the profit.</p>

<b>Define the Gurobi model "Value-Com-Electronics"</b>

In [2]:
model = Model('Value-Com-Electronics')

Set parameter Username


<b>Add Variables to the model</b>

In [3]:
card1 = model.addVar(name='HyperLink')
card2 = model.addVar(name='FastLink')
card3 = model.addVar(name='SpeedLink')
card4 = model.addVar(name='MicroLink')
card5 = model.addVar(name='EtherLink')

<b>Set the objective function to get maximal profit.</b>

In [4]:
model.setObjective(53*card1 + 48*card2 + 33*card3 + 32*card4 + 38*card5, GRB.MAXIMIZE)

<b>Add given linear constraints to the model. Since this is a linear programming<br>
exercise, we will only consider linear constraints.</b>

In [5]:
model.addConstr(card1 >= 500, name='minimun no of HyperLink cards: ')
model.addConstr(card2 >= 500, name='minimun no of FastLink cards: ')
model.addConstr(card3 >= 500, name='minimun no of SpeedLink cards: ')
model.addConstr(card4 >= 500, name='minimun no of MicroLink cards: ')
model.addConstr(card5 >= 500, name='minimun no of EtherLink cards: ')
model.addConstr(card2 == 2*card1, name='Produce twice FastLink as HyperLink ')
model.addConstr(20*card1 + 15*card2 + 10*card3 + 8*card4 + 5*card5 <= 80000
, name='minimun amount of pcb : ')
model.addConstr(28*card1 + 24*card2 + 18*card3 + 12*card4 + 16*card5 <= 100000
, name='minimun amount of resistors : ')
model.addConstr(8*card1 + 8*card2 + 4*card3 + 4*card4 + 6*card5 <= 30000
, name='minimun amount of memory chips : ')
model.addConstr(0.75*card1 + 0.6*card2 + 0.5*card3 + 0.65*card4 + 1*card5 <= 5000
, name='minimun amount of labor hours : ')


<gurobi.Constr *Awaiting Model Update*>

<b>Save the model to a file named "Value-Com-Electronics.lp"</b>

In [6]:
model.write('Value-Com-Electronics.lp')

<b>Use Gurobi optimizers to get the optimal solution</b>  

In [7]:
model.optimize()

Gurobi Optimizer version 10.0.1 build v10.0.1rc0 (win64)

CPU model: Intel(R) Core(TM) i5-1035G7 CPU @ 1.20GHz, instruction set [SSE2|AVX|AVX2|AVX512]
Thread count: 4 physical cores, 8 logical processors, using up to 8 threads

Optimize a model with 10 rows, 5 columns and 27 nonzeros
Model fingerprint: 0x2d5a1569
Coefficient statistics:
  Matrix range     [5e-01, 3e+01]
  Objective range  [3e+01, 5e+01]
  Bounds range     [0e+00, 0e+00]
  RHS range        [5e+02, 1e+05]
Presolve removed 6 rows and 1 columns
Presolve time: 0.02s
Presolved: 4 rows, 4 columns, 16 nonzeros

Iteration    Objective       Primal Inf.    Dual Inf.      Time
       0    4.5660000e+05   1.044779e+04   0.000000e+00      0s
       4    2.1500000e+05   0.000000e+00   0.000000e+00      0s

Solved in 4 iterations and 0.02 seconds (0.00 work units)
Optimal objective  2.150000000e+05


<b>Print the Results</b>

In [9]:
for v in model.getVars():
    print(v.varName, v.x)

HyperLink 500.0
FastLink 1000.0
SpeedLink 1500.0
MicroLink 2250.0
EtherLink 500.0


<b>Summary</b>

<p>Value Com can produce 500 HyperLink, 1000 FastLink<br>
1500 SpeedLink, 2250 MicroLink and 500 EtherLink cards<br>
to maximize its profits</p>

<b>Part 2: Rent-a-dent company Transportation Cost Minimization problem</b>

<p>Description: The company allows its customers to pick a rental car<br>
from one location and return it to any of its locations. Given the cost<br>
of transportation data between each of its locations. Find a optimal <br>
solution to have atleast 5 cars at location 3,4,5 and 6 while minimizing<br>
cost of transportation.</p>

In [20]:
rent = Model('Rent-a-Dent')

<b>Add Variables to the rent model</b>

<p>Here, xij represents no of cars sent from<br>
location i to location j</p>

In [21]:
x13 = rent.addVar(name="No. of cars sent from loc1 to loc3: ")
x14 = rent.addVar(name="No. of cars sent from loc1 to loc4: ")
x15 = rent.addVar(name="No. of cars sent from loc1 to loc5: ")
x16 = rent.addVar(name="No. of cars sent from loc1 to loc6: ")
x23 = rent.addVar(name="No. of cars sent from loc2 to loc3: ")
x24 = rent.addVar(name="No. of cars sent from loc2 to loc4: ")
x25 = rent.addVar(name="No. of cars sent from loc2 to loc5: ")
x26 = rent.addVar(name="No. of cars sent from loc2 to loc6: ")

<b>Set the Objective function for the model</b>

In [22]:
rent.setObjective(54*x13 + 17*x14 + 23*x15 + 30*x16 + 24*x23 + 18*x24 + 19*x25 + 31*x26, GRB.MINIMIZE)

<b>Add the constraints given in the problem statement</b>

In [28]:
rent.addConstr(x13 + x14 + x15 + x16 == 16, name='No. of Surplus cars at location 1 ')
rent.addConstr(x23 + x24 + x25 + x26 == 18, name='No. of Surplus cars at location 2 ')
rent.addConstr(x13 + x23 >= 5 , name='Min. Cars sent to location 1 ')
rent.addConstr(x13 + x23 <= 10 , name='Max. Cars sent to location 1  ')
rent.addConstr(x14 + x24 >= 5 , name='Min. Cars sent to location 2 ')
rent.addConstr(x14 + x24 >= 10 , name='Max. Cars sent to location 2 ')
rent.addConstr(x15 + x25 >= 5 , name='Min. Cars sent to location 3 ')
rent.addConstr(x15 + x25 <= 10 , name='Max. Cars sent to location 3 ')
rent.addConstr(x16 + x26 >= 5 , name='Min. Cars sent to location 4 ')
rent.addConstr(x16 + x26 <= 10 , name='Max. Cars sent to location 4 ')

<gurobi.Constr *Awaiting Model Update*>

<b>Save the model to a file named "rent-a-dent.lp"</b>

In [29]:
rent.write('rent-a-dent.lp')

<b>Perform Optimizations</b>

In [30]:
rent.optimize()

Gurobi Optimizer version 10.0.1 build v10.0.1rc0 (win64)

CPU model: Intel(R) Core(TM) i5-1035G7 CPU @ 1.20GHz, instruction set [SSE2|AVX|AVX2|AVX512]
Thread count: 4 physical cores, 8 logical processors, using up to 8 threads

Optimize a model with 23 rows, 8 columns and 68 nonzeros
Model fingerprint: 0x87174973
Coefficient statistics:
  Matrix range     [1e+00, 1e+00]
  Objective range  [2e+01, 5e+01]
  Bounds range     [0e+00, 0e+00]
  RHS range        [5e+00, 2e+01]
Presolve removed 17 rows and 0 columns
Presolve time: 0.01s
Presolved: 6 rows, 11 columns, 19 nonzeros

Iteration    Objective       Primal Inf.    Dual Inf.      Time
       0    0.0000000e+00   2.950000e+01   0.000000e+00      0s
       5    6.9600000e+02   0.000000e+00   0.000000e+00      0s

Solved in 5 iterations and 0.01 seconds (0.00 work units)
Optimal objective  6.960000000e+02


<b>Print the results</b>

In [31]:
for v in rent.getVars():
    print(v.varName, v.x)

No. of cars sent from loc1 to loc3:  0.0
No. of cars sent from loc1 to loc4:  11.0
No. of cars sent from loc1 to loc5:  0.0
No. of cars sent from loc1 to loc6:  5.0
No. of cars sent from loc2 to loc3:  5.0
No. of cars sent from loc2 to loc4:  8.0
No. of cars sent from loc2 to loc5:  5.0
No. of cars sent from loc2 to loc6:  0.0


<b>Summary</b>

<p>To minimize Cost of Transportations,<br>Rent-a-dent needs to send:<br>
<li><em>11 cars from location 1 to 4<br></em></li>
<li><em>5 cars from location 1 to 6<br></em></li>
<li><em>5 cars from location 2 to 3<br></em></li>
<li><em>8 cars from location 2 to 4<br></em></li>
<li><em>5 cars from location 5 to 5<br></em></li></p>